### PR TITLE
Golden tests

### PR DESCRIPTION
**Pull request by nevrome**
So this is an implementation of #101. Looks like a lot of code with many changes, but is actually a pretty simple setup. 

Look [here](https://github.com/poseidon-framework/poseidon-hs/tree/golden_tests#golden-tests) for a general introduction.

Some of the tests may be a bit pointless and some functionality is not tested at all. It's only a starting point to get some first coverage for our complex CLI functions.

### CODE REVIEW

**Review by stschiff — COMMENTED** [review]
Wow, what in incredible amount of work. Awesome, this will make future refactoring and feature addition so much safer. Let me try it out myself on the command line, then I'll approve.

**Comment by stschiff** [comment]
OK, sorry for taking me so long. The design and code is brillant. Thanks so much! 
However, the tests currently fail, and I'm wondering whether you can reproduce this? 
Specifically, I get (just compare the forth element of the lists):

```
1) Poseidon.GoldenTestsValidateChecksums, running a CLI pipeline for a frontend test, should produce the expected output files
       expected: ["Checksums for trident CLI output","Automatically generated with: poseidon-devtools updateGoldenTests","","fef79e81f1e35653fd507b8de4a4c3df init Schiffels/POSEIDON.yml","04686d2b18f41b9137807a0d5521670d init Schiffels/Schiffels.janno","95b093eefacc1d6499afcfe89b15d56c init Schiffels/geno.txt","b69e22b6af449870242e597c74c21697 init Wang/POSEIDON.yml","51731d04de568ede6152a81e9c1420da init Wang/Wang.janno","ae66d851301f4a761b819f97ec28fa55 init Wang/Wang_2020.bed","f4412d50110ca306919f497440b0d1a5 checksumupdate Schiffels/POSEIDON.yml","4f054a4d5b9302ba0023a7c7ded8c6af validate validate1","4f054a4d5b9302ba0023a7c7ded8c6af validate validate2","4f054a4d5b9302ba0023a7c7ded8c6af validate validate3","338208621f2547bbb0eb2f8abf17db7d list list1","c27895a4e15eb46f0d4473cd37488cae list list2","73371da7bc569c259a996b8cb0988572 list list3","a9f5b04f5b40eeff3aabf1ad7bd22771 list list4","551620f20ad8b9f4e2717c54765fff74 summarise summarise1","fb074b54961035f8903f3055c0bbb24e summarise summarise2","0a3dc431b6d4f0200127eeaaf0db9fa0 survey survey1","a90879ebee3943dba0596c115b951c05 survey survey2","3feaa876c5d05fd000470c5ba840cc2d genoconvert Wang/Wang.geno","b46831b007c2d53a12b472484b7b00b4 genoconvert Wang/Wang.snp","2faf8a7b87037451d4fd5ae9cc9af460 genoconvert Wang/Wang.ind","bb08bacdf59553b52f3525f134a600d8 genoconvert Schiffels/Schiffels.bed","67c3a8780a816ba9bf1bb9636d14d277 genoconvert Schiffels/Schiffels.bim","a9fcb59cb933d8f183f3c3f6fbf2e213 genoconvert Schiffels/Schiffels.fam","814d71c65037eb20bb8a636e804c9dd9 forge ForgePac1/ForgePac1.geno","0e8a3a81039f3344ec598da10475af63 forge ForgePac1/ForgePac1.janno","e8a7d4c9ac62234f36ea2d5b4d0a17df forge ForgePac2/ForgePac2.bed","0e8a3a81039f3344ec598da10475af63 forge ForgePac2/ForgePac2.janno","53edcde25d7b66b94ea9ee306a315864 fetch 2019_Nikitin_LBK/POSEIDON.yml","4dedd209cdc4d4a986c184d70dedecd1 fetch 2019_Nikitin_LBK/Nikitin_LBK.janno","3abe2144e4f5aea3aa0b83395ba9b355 fetch 2019_Nikitin_LBK/Nikitin_LBK.fam"]
        but got: ["Checksums for trident CLI output","Automatically generated with: poseidon-devtools updateGoldenTests","","8158efd32b30e6938bad765993631bf9 init Schiffels/POSEIDON.yml","04686d2b18f41b9137807a0d5521670d init Schiffels/Schiffels.janno","95b093eefacc1d6499afcfe89b15d56c init Schiffels/geno.txt","7f4ba7ac62a06d445b1bbde6f7abccae init Wang/POSEIDON.yml","51731d04de568ede6152a81e9c1420da init Wang/Wang.janno","ae66d851301f4a761b819f97ec28fa55 init Wang/Wang_2020.bed","1e7a588c612f7f83236869ddcde008c6 checksumupdate Schiffels/POSEIDON.yml","4f054a4d5b9302ba0023a7c7ded8c6af validate validate1","4f054a4d5b9302ba0023a7c7ded8c6af validate validate2","4f054a4d5b9302ba0023a7c7ded8c6af validate validate3","338208621f2547bbb0eb2f8abf17db7d list list1","c27895a4e15eb46f0d4473cd37488cae list list2","73371da7bc569c259a996b8cb0988572 list list3","a9f5b04f5b40eeff3aabf1ad7bd22771 list list4","551620f20ad8b9f4e2717c54765fff74 summarise summarise1","fb074b54961035f8903f3055c0bbb24e summarise summarise2","0a3dc431b6d4f0200127eeaaf0db9fa0 survey survey1","a90879ebee3943dba0596c115b951c05 survey survey2","3feaa876c5d05fd000470c5ba840cc2d genoconvert Wang/Wang.geno","b46831b007c2d53a12b472484b7b00b4 genoconvert Wang/Wang.snp","2faf8a7b87037451d4fd5ae9cc9af460 genoconvert Wang/Wang.ind","bb08bacdf59553b52f3525f134a600d8 genoconvert Schiffels/Schiffels.bed","67c3a8780a816ba9bf1bb9636d14d277 genoconvert Schiffels/Schiffels.bim","a9fcb59cb933d8f183f3c3f6fbf2e213 genoconvert Schiffels/Schiffels.fam","814d71c65037eb20bb8a636e804c9dd9 forge ForgePac1/ForgePac1.geno","0e8a3a81039f3344ec598da10475af63 forge ForgePac1/ForgePac1.janno","e8a7d4c9ac62234f36ea2d5b4d0a17df forge ForgePac2/ForgePac2.bed","0e8a3a81039f3344ec598da10475af63 forge ForgePac2/ForgePac2.janno","53edcde25d7b66b94ea9ee306a315864 fetch 2019_Nikitin_LBK/POSEIDON.yml","4dedd209cdc4d4a986c184d70dedecd1 fetch 2019_Nikitin_LBK/Nikitin_LBK.janno","3abe2144e4f5aea3aa0b83395ba9b355 fetch 2019_Nikitin_LBK/Nikitin_LBK.fam"]
```

So I think a checksum doesn't match. I'll try to understand what caused that.

**Comment by stschiff** [comment]
I think this occurs only with the `init` command. Is it possible that it has to do with the fact that the resulting absolute paths in the packages are platform dependent?

**Comment by nevrome** [comment]
I think the issue are the `lastModified` fields in the POSEIDON.yml files. You could explore this by running 

```bash
poseidon-devtools updateGoldenTests -d .
git diff
```

This gives me

```
--- a/test/testDat/poseidonHSGoldenTestCheckSumFile.txt
+++ b/test/testDat/poseidonHSGoldenTestCheckSumFile.txt
@@ -1,13 +1,13 @@
 Checksums for trident CLI output
 Automatically generated with: poseidon-devtools updateGoldenTests
 
-fef79e81f1e35653fd507b8de4a4c3df init Schiffels/POSEIDON.yml
+8158efd32b30e6938bad765993631bf9 init Schiffels/POSEIDON.yml
 04686d2b18f41b9137807a0d5521670d init Schiffels/Schiffels.janno
 95b093eefacc1d6499afcfe89b15d56c init Schiffels/geno.txt
-b69e22b6af449870242e597c74c21697 init Wang/POSEIDON.yml
+7f4ba7ac62a06d445b1bbde6f7abccae init Wang/POSEIDON.yml
 51731d04de568ede6152a81e9c1420da init Wang/Wang.janno
 ae66d851301f4a761b819f97ec28fa55 init Wang/Wang_2020.bed
-f4412d50110ca306919f497440b0d1a5 checksumupdate Schiffels/POSEIDON.yml
+1e7a588c612f7f83236869ddcde008c6 checksumupdate Schiffels/POSEIDON.yml
 4f054a4d5b9302ba0023a7c7ded8c6af validate validate1
 4f054a4d5b9302ba0023a7c7ded8c6af validate validate2
 4f054a4d5b9302ba0023a7c7ded8c6af validate validate3
diff --git a/test/testDat/poseidonHSGoldenTestData/ForgePac1/POSEIDON.yml b/test/testDat/poseidonHSGoldenTestData/ForgePac1/POSEIDON.yml
index 7df7f94..cc79c3f 100644
--- a/test/testDat/poseidonHSGoldenTestData/ForgePac1/POSEIDON.yml
+++ b/test/testDat/poseidonHSGoldenTestData/ForgePac1/POSEIDON.yml
@@ -5,7 +5,7 @@ contributor:
 - name: John Doe
   email: john@doe.net
 packageVersion: 0.1.0
-lastModified: 2021-06-05
+lastModified: 2021-07-06
 genotypeData:
   format: EIGENSTRAT
   genoFile: ForgePac1.geno
diff --git a/test/testDat/poseidonHSGoldenTestData/ForgePac2/POSEIDON.yml b/test/testDat/poseidonHSGoldenTestData/ForgePac2/POSEIDON.yml
index ae9d643..8ca43d5 100644
--- a/test/testDat/poseidonHSGoldenTestData/ForgePac2/POSEIDON.yml
+++ b/test/testDat/poseidonHSGoldenTestData/ForgePac2/POSEIDON.yml
@@ -5,7 +5,7 @@ contributor:
 - name: John Doe
   email: john@doe.net
 packageVersion: 0.1.0
-lastModified: 2021-06-05
+lastModified: 2021-07-06
 genotypeData:
   format: PLINK
   genoFile: ForgePac2.bed
```

**Comment by nevrome** [comment]
This is an issue I obviously missed on the day I prepared this PR :smile:

How can we fix this? Either we give `init` and `update` (#111) an option to make this date static for our purposes or we exclude this row somehow from the checksum comparison in the golden tests.

**Comment by stschiff** [comment]
Ah yes, of course. OK, I think I would definitely lean towards solving this within `testPipelineInit` and `testPipelineChecksumupdate`. It might be more cumbersome, but in the end it's less elegant to put in an option into the core init and update functions just for the test.

**Comment by stschiff** [comment]
Happy to take a stab at that.

**Comment by stschiff** [comment]
OK. I think I may have fixed it. Have a look.

File: test/Poseidon/GoldenTestsRunCommands.hs [hunk 1]
```diff
@@ -96,7 +99,8 @@ testPipelineInit testDir checkFilePath testPacsDir = do
         , _initPacPath   = testDir </> "Schiffels"
         , _initPacName   = "Schiffels"
     }
-    runAndChecksumFiles checkFilePath testDir (runInit initOpts1) "init" [
+    let action = runInit initOpts1 >> patchLastModified testDir ("Schiffels" </> "POSEIDON.yml")
```
Comment (nevrome) [inline test/Poseidon/GoldenTestsRunCommands.hs:102]: Wow! What a brilliant operator!

File: test/Poseidon/GoldenTestsRunCommands.hs [hunk 2]
```diff
@@ -110,12 +114,22 @@ testPipelineInit testDir checkFilePath testPacsDir = do
...
+    let action2 = runInit initOpts2 >> patchLastModified testDir ("Wang" </> "POSEIDON.yml")
+    runAndChecksumFiles checkFilePath testDir action2 "init" [
           "Wang" </> "POSEIDON.yml"
         , "Wang" </> "Wang.janno"
         , "Wang" </> "Wang_2020.bed"
         ]
 
+patchLastModified :: FilePath -> FilePath -> IO ()
+patchLastModified testDir yamlFile = do
+    lines_ <- T.lines <$> T.readFile (testDir </> yamlFile)
```
Comment (nevrome) [inline test/Poseidon/GoldenTestsRunCommands.hs:126]: Isn't there an issue with laziness here? Reading and overwriting causes issues sometimes, when done with readFile and writeFile, right? Is Text.readFile strict?

Comment (stschiff) [reply test/Poseidon/GoldenTestsRunCommands.hs:126]: Yes it is in Data.Text.IO. I had used the lazy String version before from Prelude and promptly stumbled into problems. That's why I used the one from Data.Text.IO

**Comment by stschiff** [comment]
Alright. One day has passed and the tests still pass :-)

Good to merge this PR from my end. I'll leave that to you, @nevrome. I'll deal with the other PR now (#111)

File: test/Poseidon/GoldenTestsRunCommands.hs (hunk 2 above)
Comment (nevrome) [reply test/Poseidon/GoldenTestsRunCommands.hs:126]: Good to know. I solved this issue recently by deleting the file before rewriting it. Probably an explicitly strict solution is less of a hack...

https://github.com/poseidon-framework/poseidon-hs/blob/6652c6b2f1b44f0c9a09f084e66d39495454e4c5/src/Poseidon/CLI/Update.hs#L83-L86

**Pull request merged**